dpctl/_sycl_platform.pyx

```diff
@@ -366,6 +375,78 @@ cdef class SyclPlatform(_SyclPlatform):
         """
         return DPCTLPlatform_Hash(self._platform_ref)
 
+    def get_devices(self, device_type=device_type_t.all):
+        """
+        Returns the list of :class:`dpctl.SyclDevice` objects associated with
+        :class:`dpctl.SyclPlatform` instance selected based on
+        the given :class:`dpctl.device_type`.
+
+        Args:
+            device_type (optional):
+                A :class:`dpctl.device_type` enum value or a string that
+                specifies a SYCL device type. Currently, accepted values are:
+                "gpu", "cpu", "accelerator", or "all".
+                Default: ``dpctl.device_type.all``.
+
+        Returns:
+            list:
+                A :obj:`list` of :class:`dpctl.SyclDevice` objects
+                that belong to this platform.
+
+        Raises:
+            TypeError:
+                If `device_type` is not a str or :class:`dpctl.device_type`
+                enum.
+            ValueError:
+                If the ``DPCTLPlatform_GetDevices`` call returned
+                ``NULL`` instead of a ``DPCTLDeviceVectorRef`` object.
+        """
+        cdef _device_type DTy = _device_type._ALL_DEVICES
+        cdef DPCTLDeviceVectorRef DVRef = NULL
+        cdef size_t num_devs
+        cdef size_t i
+        cdef DPCTLSyclDeviceRef DRef
+
+        if isinstance(device_type, str):
+            dty_str = device_type.strip().lower()
+            if dty_str == "accelerator":
+                DTy = _device_type._ACCELERATOR
+            elif dty_str == "all":
+                DTy = _device_type._ALL_DEVICES
+            elif dty_str == "cpu":
+                DTy = _device_type._CPU
+            elif dty_str == "gpu":
+                DTy = _device_type._GPU
+            else:
+                DTy = _device_type._UNKNOWN_DEVICE
+        elif isinstance(device_type, device_type_t):
+            if device_type == device_type_t.all:
+                DTy = _device_type._ALL_DEVICES
+            elif device_type == device_type_t.accelerator:
+                DTy = _device_type._ACCELERATOR
+            elif device_type == device_type_t.cpu:
+                DTy = _device_type._CPU
+            elif device_type == device_type_t.gpu:
+                DTy = _device_type._GPU
+            else:
+                DTy = _device_type._UNKNOWN_DEVICE
+        else:
+            raise TypeError(
+                "device type should be specified as a str or an "
+                "``enum_types.device_type``."
+            )
+        DVRef = DPCTLPlatform_GetDevices(self.get_platform_ref(), DTy)
+        if (DVRef is NULL):
+            raise ValueError("Internal error: NULL device vector encountered")
+        num_devs = DPCTLDeviceVector_Size(DVRef)
+        devices = []
+        for i in range(num_devs):
+            DRef = DPCTLDeviceVector_GetAt(DVRef, i)
+            devices.append(SyclDevice._create(DRef))
```

Comment:
It seems exactly the same code present in many places, somewhere it is declared as internal function:
```cython
cdef list _get_devices(DPCTLDeviceVectorRef DVRef):
    cdef list devices = []
    cdef size_t nelems = 0
    if DVRef:
        nelems = DPCTLDeviceVector_Size(DVRef)
        for i in range(0, nelems):
            DRef = DPCTLDeviceVector_GetAt(DVRef, i)
            D = SyclDevice._create(DRef)
            devices.append(D)

    return devices
```
Should it be exposed as internal helper function and reused in all that places?

Reply:
Yes, we've reused this a few times, so I agree with that. Perhaps a `_common.pyx` or `_helper.pyx` is a sensible inclusion.

This may be better saved for a follow-up PR though.